dev-manuals/backend/testing.md

```diff
@@ -1,4 +1,162 @@
+
+We use [Mockito](https://site.mockito.org/) for mocking. It allows us to mock classes and interfaces and to verify
+that certain methods were called on the mock.
+
+### Naming conventions
+
+There are no strict naming conventions for unit tests. However, the test method should describe what is tested and
+what the expected result is. For example, if we test the delete method of the `UserService`, we could name the test
+method `testDeleteExistingUser()`.
+Other approaches put the expected result in the test method name, for example `deleteUser_UserExists_UserDeleted()`,
+but this often leads to very long test method names.
+
+We therefore prefer to put the expected result in the JavaDoc of the test method:
+The JavaDoc of the test method should describe what is tested and what the expected result is, after the given
+pattern: `Given <some context>, when <some action is taken>, then <some result is expected>`.
+
+### Example
+
+```java
+public class ChapterServiceTest {
+    /**
+     * Given a valid ChapterId
+     * When deleteChapter is called
+     * Then the UUID is returned and the delete method of the repository is called
+     */
+    @Test
+    public void testDeleteChapterSuccessful() {
+        // arrange test data
+        UUID testChapterId = UUID.randomUUID();
+
+        // mock repository
+        doNothing()
+                .when(chapterRepository).deleteById(any());
+        doReturn(true)
+                .when(chapterRepository).existsById(any());
+
+        // act
+        UUID deletedChapterId = chapterService.deleteChapter(testChapterId);
+
+        // assert
+        assertThat(deletedChapterId, is(testChapterId));
+
+        // verify that the repository was called
+        verify(chapterRepository).deleteById(testChapterId);
+    }
+}
+```
+
+## Integration/API testing
+
+As our microservices are rather small, we do API testing instead of what is usually referred to as integration testing.
+This means that we test the graphQL API of the microservices instead of interaction of classes in the microservice.
+Having integration tests in addition to unit tests and API tests would be redundant, as the API tests already test
+the interaction of the classes and the database.
+
+Of course, if it makes sense to have integration tests in addition in some cases, we can always add them.
+
+### What to test
+
+We test the graphQL API of the microservices. This means that we test the graphQL queries and mutations that are
+exposed by the microservice. As in unit tests, think of all special cases and exceptional paths that should be tested.
+
+If a functionality is tests in detail in a unit test, we do not have to test it again in that much detail in an API
+test. But in case of doubt, we prefer to have more tests than less.
+
+### How to test
+
+We provide the `GraphQlApiTest` annotation that sets up the test environment for API tests. 
+It starts the spring application context and requires the database to be running. 
+
```

Comment:
For integration testing, we could use an H2 database that is automatically started and stopped. I used it in the content_service to try it out and it worked fine.

Reply:
I'm not sure if using a separate database type for integration tests is a good idea. Every database technology like Postgres and MySQL have their own caveats and we might miss some errors that would occur in production, but not in the in memory database.

Reply:
Here is stackoverflow post on this: https://stackoverflow.com/questions/68180468/is-it-bad-practice-to-use-h2db-for-integration-test-in-a-spring-boot-context

Reply:
As my computer is already struggling, it was very useful to use H2 database. Therefore I would like to proceed using H2 in the content service at the moment, but you are right, we do not need this in the test strategy.

Reply:
Do you have an idea how to this can be done? So maybe we have a different profile for local running or something?

Reply:
When you set an environment variable SPRING_CONFIG_NAME=dev , then Spring Boot will read properties from application-dev.properties . In this file, I added the settings for H2 in the content service.

Reply:
I add this to the template-repo pull request

Reply:
Okay that is actually more difficult than expected. Would it be okay to leave it as it is for now and work on a dev environment setup next sprint?